tools/windows/install-help/cal/PropertyReplacer.cpp

```diff
@@ -0,0 +1,210 @@
+#include "stdafx.h"
+#include "PropertyReplacer.h"
+
+namespace
+{
+    template <class Map>
+    bool has_key(Map const &m, const typename Map::key_type &key)
+    {
+        auto const &it = m.find(key);
+        return it != m.end();
+    }
+
+    bool to_bool(std::wstring str)
+    {
+        std::transform(str.begin(), str.end(), str.begin(), ::tolower);
+        std::wistringstream is(str);
+        bool b;
+        is >> std::boolalpha >> b;
+        return b;
+    }
+
+    typedef std::function<std::wstring(std::wstring const &, const property_retriever &)> formatter_func;
+
+    /// <summary>
+    /// Simply concatenates <paramref name="str"/> with the value of the matching property.
+    /// </summary>
+    /// <param name="str">The string to use as a replacement.</param>
+    /// <returns>A function that conforms to <see cref="formatter_func"> that when called with a property value,
+    /// will return a concatenated string of <paramref name="str"/> and the property value. </returns>
+    formatter_func format_simple_value(const std::wstring &str)
+    {
+        return [str](std::wstring const &propertyValue, const property_retriever &)
+        {
+            return str + propertyValue;
+        };
+    }
+
+    formatter_func simple_replace(const std::wstring &str)
+    {
+        return [str](std::wstring const &, const property_retriever &) { return str; };
+    }
+
+    std::wstring format_tags(const std::wstring &tags, const property_retriever &)
+    {
+        std::wistringstream valueStream(tags);
+        std::wstringstream result;
+        std::wstring token;
+        result << L"tags: ";
+        while (std::getline(valueStream, token, static_cast<wchar_t>(',')))
+        {
+            result << std::endl << L"  - " << token;
+        }
+        return result.str();
+    };
+
+    std::wstring format_proxy(std::wstring proxyHost, const property_retriever &propertyRetriever)
+    {
+        const auto proxyPort = propertyRetriever(L"PROXY_PORT");
+        const auto proxyUser = propertyRetriever(L"PROXY_USER");
+        const auto proxyPassword = propertyRetriever(L"PROXY_PASSWORD");
+        std::wstringstream proxy;
+        std::size_t schemeEnd = proxyHost.find(L"://", 0);
+        if (schemeEnd == std::string::npos)
+        {
+            proxy << "http://";
+        }
+        else
+        {
+            proxy << proxyHost.substr(0, schemeEnd+3);
+            proxyHost.erase(0, schemeEnd+3);
+        }
+        if (proxyUser)
+        {
+            proxy << *proxyUser;
+            if (proxyPassword)
+            {
+                proxy << L":" << *proxyPassword;
```

Comment:
Not a blocker since the old code behaved the same but I think in the future we need to ensure that we escape both username and password strings since both can include `:` and/or `@` symbols.